File: python/src/mshzip/unpacker.py
'MSH format restore (unpack) - frame parsing + dict accumulation'
from __future__ import annotations

import gzip
import math
import struct

from . import varint
from .constants import (
    MAGIC, Codec, Flag, KNOWN_FLAGS,
    FRAME_HEADER_SIZE,
    BITDICT_EXTRA_HEADER_SIZE,
    COORDDICT_EXTRA_HEADER_SIZE,
)
from .dict_store import DictStore
from .bit_reader import write_all_chunks


class Unpacker:
    'Restore MSH format data to original.'

    def __init__(
        self,
        dict_store: DictStore | None = None,
        dict_dir: str | None = None,
    ) -> None:
        self.dict: list[bytes] = []
        self._dict_store = dict_store
        self._dict_dir = dict_dir
        self._coord_dict_unpacker = None

    def close(self) -> None:
        """리소스 정리."""
        if self._coord_dict_unpacker is not None:
            self._coord_dict_unpacker.close()
            self._coord_dict_unpacker = None

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

    def unpack(self, data: bytes | bytearray | memoryview) -> bytes:
        'Restore MSH data to original.'
        if isinstance(data, memoryview):
            buf = data
        else:
            buf = memoryview(data) if data else memoryview(b'')
        parts: list[bytes] = []
        offset = 0

        while offset < len(buf):
            restored, bytes_consumed = self._read_frame(buf, offset)
            parts.append(restored)
            offset += bytes_consumed

        return b''.join(parts)

    def _read_frame(
        self,
        buf: memoryview | bytes | bytearray,
        offset: int,
    ) -> tuple[bytes, int]:
        'Read and restore single frame. Returns (data, bytes_consumed).'
        start_offset = offset

        if offset + FRAME_HEADER_SIZE > len(buf):
            raise ValueError(f'Insufficient frame header: offset={offset}')

        # Verify magic number
        magic = bytes(buf[offset:offset + 4])
        if magic != MAGIC:
            raise ValueError(f'Invalid magic number: {magic!r}')

        # Header parsing (manual offset - 1:1 mapping with Node.js original)
        off = offset + 4  # after magic
        version = struct.unpack_from('<H', buf, off)[0]; off += 2
        flags = struct.unpack_from('<H', buf, off)[0]; off += 2

        # 알 수 없는 flags 방어적 에러 (하위 호환성)
        unknown_flags = flags & ~KNOWN_FLAGS
        if unknown_flags:
            raise ValueError(
                f'Unsupported flags: 0x{unknown_flags:04x}. '
                'This file requires a newer version of mshzip.'
            )

        chunk_size = struct.unpack_from('<I', buf, off)[0]; off += 4
        codec_id = buf[off]; off += 1
        off += 3  # padding
        orig_bytes_lo = struct.unpack_from('<I', buf, off)[0]; off += 4
        orig_bytes_hi = struct.unpack_from('<I', buf, off)[0]; off += 4
        orig_bytes = orig_bytes_hi * 0x100000000 + orig_bytes_lo
        dict_entries = struct.unpack_from('<I', buf, off)[0]; off += 4
        seq_count = struct.unpack_from('<I', buf, off)[0]; off += 4

        has_hier_dedup = (flags & Flag.HIERDEDUP) != 0
        has_external_dict = (flags & Flag.EXTERNAL_DICT) != 0
        has_bit_dict = (flags & Flag.BITDICT) != 0
        has_coord_dict = (flags & Flag.COORDDICT) != 0

        # COORDDICT 상호 배타 검증
        if has_coord_dict and (flags & (Flag.BITDICT | Flag.HIERDEDUP | Flag.MULTILEVEL | Flag.EXTERNAL_DICT)):
            raise ValueError('COORDDICT는 다른 모드와 동시 사용 불가')

        # BITDICT와 HIERDEDUP/EXTERNAL_DICT 상호 배타 검증
        if has_bit_dict and (flags & (Flag.HIERDEDUP | Flag.EXTERNAL_DICT)):
            raise ValueError('BITDICT는 HIERDEDUP/EXTERNAL_DICT와 동시 사용 불가')

        # COORDDICT: Extra Header 읽기
        coord_dimensions = 0
        coord_rs_axes = 0
        if has_coord_dict:
            coord_dimensions = struct.unpack_from('<H', buf, off)[0]; off += 2
            off += 2  # bitsPerAxis skip
            off += 1  # hammingBits skip
            coord_rs_axes = buf[off]; off += 1
            off += 2  # reserved

        # BITDICT: bitDepth 읽기
        bit_depth = 0
        if has_bit_dict:
            bit_depth = struct.unpack_from('<H', buf, off)[0]; off += BITDICT_EXTRA_HEADER_SIZE

        # EXTERNAL_DICT: baseDictCount 읽기 + 외부 딕셔너리 로드
        base_dict_count = 0
        if has_external_dict:
            base_dict_count = struct.unpack_from('<I', buf, off)[0]; off += 4
            self._load_external_dict(chunk_size, base_dict_count)

        # Compressed payload size
        payload_size = struct.unpack_from('<I', buf, off)[0]; off += 4

        # Read compressed payload
        compressed_payload = bytes(buf[off:off + payload_size])
        off += payload_size

        # CRC32 check
        has_crc = (flags & Flag.CRC32) != 0
        if has_crc:
            off += 4  # CRC skip (Node.js original also skips verification)

        # Decompress payload
        raw_payload = self._decompress(compressed_payload, codec_id)

        # ── COORDDICT 모드 ──
        if has_coord_dict:
            if seq_count > 0 and orig_bytes > 0:
                if self._coord_dict_unpacker is None:
                    from .coord_dict import CoordDictUnpacker
                    self._coord_dict_unpacker = CoordDictUnpacker()
                restored_data = self._coord_dict_unpacker.decode(
                    raw_payload, coord_dimensions, coord_rs_axes, seq_count, orig_bytes
                )
            else:
                restored_data = b''
            return (restored_data, off - start_offset)

        # ── BITDICT 모드: 사전 섹션 없이 바로 시퀀스 복원 ──
        if has_bit_dict:
            if seq_count > 0 and orig_bytes > 0:
                indices, _ = varint.decode_array(raw_payload, 0, seq_count)
                restored = write_all_chunks(indices, bit_depth)
                restored_data = restored[:orig_bytes]
            else:
                restored_data = b''
            return (restored_data, off - start_offset)

        # Parse dict section
        payload_off = 0

        if has_hier_dedup and dict_entries > 0:
            # ── 계층적 Dedup 복원: Dict2 + Seq2 → Dict1 ──

            # Hier Header (8B)
            sub_chunk_size = struct.unpack_from('<I', raw_payload, payload_off)[0]
            payload_off += 4
            dict2_entries = struct.unpack_from('<I', raw_payload, payload_off)[0]
            payload_off += 4

            # Dict2 Section
            dict2: list[bytes] = []
            for _ in range(dict2_entries):
                dict2.append(bytes(raw_payload[payload_off:payload_off + sub_chunk_size]))
                payload_off += sub_chunk_size

            # Seq2 Section → Dict1 복원
            sub_chunks_per_chunk = math.ceil(chunk_size / sub_chunk_size)
            seq2_count = dict_entries * sub_chunks_per_chunk
            seq2_indices, seq2_bytes = varint.decode_array(
                raw_payload, payload_off, seq2_count
            )
            payload_off += seq2_bytes

            # 서브청크 재조립 → Dict1 복원
            for i in range(dict_entries):
                parts: list[bytes] = []
                for j in range(sub_chunks_per_chunk):
                    idx = seq2_indices[i * sub_chunks_per_chunk + j]
                    if idx >= len(dict2):
                        raise ValueError(
                            f'Dict2 index out of range: {idx} >= {len(dict2)}'
                        )
                    parts.append(dict2[idx])
                restored = b''.join(parts)[:chunk_size]
                self.dict.append(restored)
        else:
            # ── 기존 복원 ──
            for _ in range(dict_entries):
                chunk = bytes(raw_payload[payload_off:payload_off + chunk_size])
                self.dict.append(chunk)
                payload_off += chunk_size

        # Parse sequence section
        if seq_count > 0 and orig_bytes > 0:
            indices, _ = varint.decode_array(raw_payload, payload_off, seq_count)

            chunks: list[bytes] = []
            for idx in indices:
                if idx >= len(self.dict):
                    raise ValueError(
                        f'Dict index out of range: {idx} >= {len(self.dict)}'
                    )
                chunks.append(self.dict[idx])

            full_data = b''.join(chunks)
            restored_data = full_data[:orig_bytes]
        else:
            restored_data = b''

        return (restored_data, off - start_offset)

    def _load_external_dict(self, chunk_size: int, base_dict_count: int) -> None:
        '외부 딕셔너리 로드 (EXTERNAL_DICT 프레임 처리 시).'
        if len(self.dict) >= base_dict_count:
            return

        store = self._dict_store or DictStore(dict_dir=self._dict_dir)
        loaded = store.load(chunk_size)

        if loaded['entry_count'] < base_dict_count:
            raise ValueError(
                f'외부 딕셔너리 엔트리 부족: {loaded["entry_count"]} < {base_dict_count}. '
                f'dict-{chunk_size}.mshdict 파일이 손상되었거나 버전이 불일치합니다.'
            )

        self.dict = loaded['dict_chunks'][:base_dict_count]

    @staticmethod
    def _decompress(data: bytes, codec_id: int) -> bytes:
        'Decompress payload.'
        if not data:
            return data

        if codec_id == Codec.NONE:
            return data
        elif codec_id == Codec.GZIP:
            return gzip.decompress(data)
        else:
            raise ValueError(f'Unsupported codec ID: {codec_id}')

File: nodejs/lib/unpacker.js
'use strict';

const zlib = require('zlib');
const {
  MAGIC, CODEC, FLAG, KNOWN_FLAGS,
  FRAME_HEADER_SIZE,
  BITDICT_EXTRA_HEADER_SIZE,
  COORDDICT_EXTRA_HEADER_SIZE,
} = require('./constants');
const varint = require('./varint');
const { DictStore } = require('./dict-store');
const { writeAllChunks } = require('./bit-reader');

/**
 * Restore MSH format data to original (unpack)
 *
 * Parse frame by frame:
 * 1. Read frame header
 * 2. Decompress payload
 * 3. Restore new chunks from dict section -> accumulate in global dictionary
 * 4. Restore original chunk order using sequence section indices
 * 5. Trim last chunk using origBytes
 */

class Unpacker {
  /**
   * @param {Object} [opts]
   * @param {DictStore} [opts.dictStore] - 외부 딕셔너리 (DictStore 인스턴스)
   * @param {string} [opts.dictDir] - DictStore 디렉토리
   */
  constructor(opts = {}) {
    // Global dictionary: accumulated across frames
    this.dict = [];
    this._dictStore = opts.dictStore || null;
    this._dictDir = opts.dictDir || null;
  }

  /**
   * Restore MSH data to original
   * @param {Buffer} input - MSH compressed data
   * @returns {Buffer} - Restored original data
   */
  unpack(input) {
    const parts = [];
    let offset = 0;

    while (offset < input.length) {
      const { data, bytesConsumed } = this._readFrame(input, offset);
      parts.push(data);
      offset += bytesConsumed;
    }

    return Buffer.concat(parts);
  }

  /**
   * Read and restore a single frame
   */
  _readFrame(input, offset) {
    const startOffset = offset;

    // Verify magic number
    if (offset + FRAME_HEADER_SIZE > input.length) {
      throw new Error(`Insufficient frame header: offset=${offset}`);
    }

    const magic = input.slice(offset, offset + 4);
    if (!magic.equals(MAGIC)) {
      throw new Error(`Invalid magic number: ${magic.toString()}`);
    }

    // Parse header
    let off = offset;
    off += 4; // magic
    const version = input.readUInt16LE(off); off += 2;
    const flags = input.readUInt16LE(off); off += 2;

    // 알 수 없는 flags 방어적 에러 (하위 호환성)
    const unknownFlags = flags & ~KNOWN_FLAGS;
    if (unknownFlags) {
      throw new Error(
        `Unsupported flags: 0x${unknownFlags.toString(16)}. ` +
        'This file requires a newer version of mshzip.'
      );
    }

    const chunkSize = input.readUInt32LE(off); off += 4;
    const codecId = input.readUInt8(off); off += 1;
    off += 3; // padding
    const origBytesLo = input.readUInt32LE(off); off += 4;
    const origBytesHi = input.readUInt32LE(off); off += 4;
    const origBytes = origBytesHi * 0x100000000 + origBytesLo;
    const dictEntries = input.readUInt32LE(off); off += 4;
    const seqCount = input.readUInt32LE(off); off += 4;

    const hasHierDedup = (flags & FLAG.HIERDEDUP) !== 0;
    const hasExternalDict = (flags & FLAG.EXTERNAL_DICT) !== 0;
    const hasBitDict = (flags & FLAG.BITDICT) !== 0;
    const hasCoordDict = (flags & FLAG.COORDDICT) !== 0;

    // COORDDICT 상호 배타 검증
    if (hasCoordDict && (flags & (FLAG.BITDICT | FLAG.HIERDEDUP | FLAG.MULTILEVEL | FLAG.EXTERNAL_DICT))) {
      throw new Error('COORDDICT는 다른 모드와 동시 사용 불가');
    }

    // BITDICT와 HIERDEDUP/EXTERNAL_DICT 상호 배타 검증
    if (hasBitDict && (flags & (FLAG.HIERDEDUP | FLAG.EXTERNAL_DICT))) {
      throw new Error('BITDICT는 HIERDEDUP/EXTERNAL_DICT와 동시 사용 불가');
    }

    // COORDDICT: Extra Header 읽기
    let coordDimensions = 0, coordRsAxes = 0;
    if (hasCoordDict) {
      coordDimensions = input.readUInt16LE(off); off += 2; // dimensions
      off += 2; // bitsPerAxis (skip)
      off += 1; // hammingBits (skip)
      coordRsAxes = input.readUInt8(off); off += 1;        // rsAxes
      off += 2; // reserved
    }

    // BITDICT: bitDepth 읽기
    let bitDepth = 0;
    if (hasBitDict) {
      bitDepth = input.readUInt16LE(off); off += BITDICT_EXTRA_HEADER_SIZE;
    }

    // EXTERNAL_DICT: baseDictCount 읽기 + 외부 딕셔너리 로드
    let baseDictCount = 0;
    if (hasExternalDict) {
      baseDictCount = input.readUInt32LE(off); off += 4;
      this._loadExternalDict(chunkSize, baseDictCount);
    }

    // Compressed payload size
    const payloadSize = input.readUInt32LE(off); off += 4;

    // Read compressed payload
    const compressedPayload = input.slice(off, off + payloadSize);
    off += payloadSize;

    // CRC32 check
    const hasCRC = (flags & FLAG.CRC32) !== 0;
    if (hasCRC) {
      const storedCRC = input.readUInt32LE(off);
      off += 4;
      // CRC verification is optional (performance consideration)
    }

    // Decompress payload
    const rawPayload = this._decompress(compressedPayload, codecId);

    // ── COORDDICT 모드 ──
    if (hasCoordDict) {
      let data;
      if (seqCount > 0 && origBytes > 0) {
        const { CoordDictUnpacker } = require('./coord-dict');
        const cdu = new CoordDictUnpacker();
        data = cdu.decode(rawPayload, coordDimensions, coordRsAxes, seqCount, origBytes);
      } else {
        data = Buffer.alloc(0);
      }
      return { data, bytesConsumed: off - startOffset };
    }

    // ── BITDICT 모드: 사전 섹션 없이 바로 시퀀스 복원 ──
    if (hasBitDict) {
      let data;
      if (seqCount > 0 && origBytes > 0) {
        const { values: indices } = varint.decodeArray(rawPayload, 0, seqCount);
        const restored = writeAllChunks(indices, bitDepth);
        data = restored.slice(0, origBytes);
      } else {
        data = Buffer.alloc(0);
      }
      return { data, bytesConsumed: off - startOffset };
    }

    // Parse dict section
    let payloadOff = 0;

    if (hasHierDedup && dictEntries > 0) {
      const hasMultilevel = (flags & FLAG.MULTILEVEL) !== 0;

      if (hasMultilevel) {
        // ── N단계 계층적 Dedup 복원 ──
        const hierLevels = rawPayload.readUInt8(payloadOff); payloadOff += 1;
        payloadOff += 3; // reserved

        const levels = [];
        for (let i = 0; i < hierLevels; i++) {
          const subChunkSize_l = rawPayload.readUInt32LE(payloadOff); payloadOff += 4;
          const dictCount = rawPayload.readUInt32LE(payloadOff); payloadOff += 4;
          levels.push({ subChunkSize: subChunkSize_l, dictCount });
        }

        // 최심층 딕셔너리 읽기
        const deepest = levels[hierLevels - 1];
        let currentDict = [];
        for (let i = 0; i < deepest.dictCount; i++) {
          currentDict.push(Buffer.from(
            rawPayload.slice(payloadOff, payloadOff + deepest.subChunkSize)
          ));
          payloadOff += deepest.subChunkSize;
        }

        // 딕셔너리 역순 복원 — 플랫 버퍼 방식 (오름차순/비표준 조합 지원)
        for (let level = hierLevels - 1; level >= 0; level--) {
          let parentEntrySize, parentCount;
          if (level === 0) {
            parentEntrySize = chunkSize;
            parentCount = dictEntries;
          } else {
            parentEntrySize = levels[level - 1].subChunkSize;
            parentCount = levels[level - 1].dictCount;
          }

          const parentBufferSize = parentCount * parentEntrySize;
          const currentSubSize = levels[level].subChunkSize;
          const totalSubChunks = Math.ceil(parentBufferSize / currentSubSize);

          const { values: seqIndicesLevel, bytesRead: seqBytes } =
            varint.decodeArray(rawPayload, payloadOff, totalSubChunks);
          payloadOff += seqBytes;

          // 서브청크 → 부모 버퍼 복원
          const subChunkBufs = [];
          for (const idx of seqIndicesLevel) {
            if (idx >= currentDict.length) {
              throw new Error(
                `Dict level ${level} index out of range: ${idx} >= ${currentDict.length}`
              );
            }
            subChunkBufs.push(currentDict[idx]);
          }
          const parentBuffer = Buffer.concat(subChunkBufs).slice(0, parentBufferSize);

          // 부모 버퍼를 개별 항목으로 분할
          currentDict = [];
          for (let i = 0; i < parentCount; i++) {
            currentDict.push(Buffer.from(
              parentBuffer.slice(i * parentEntrySize, (i + 1) * parentEntrySize)
            ));
          }
        }

        // currentDict = Dict1 복원 완료
        for (const chunk of currentDict) {
          this.dict.push(chunk);
        }
      } else {
        // ── 기존 2단계 계층적 Dedup 복원 ──
        const subChunkSize = rawPayload.readUInt32LE(payloadOff); payloadOff += 4;
        const dict2Entries = rawPayload.readUInt32LE(payloadOff); payloadOff += 4;

        const dict2 = [];
        for (let i = 0; i < dict2Entries; i++) {
          dict2.push(Buffer.from(rawPayload.slice(payloadOff, payloadOff + subChunkSize)));
          payloadOff += subChunkSize;
        }

        const subChunksPerChunk = Math.ceil(chunkSize / subChunkSize);
        const seq2Count = dictEntries * subChunksPerChunk;
        const { values: seq2Indices, bytesRead: seq2Bytes } =
          varint.decodeArray(rawPayload, payloadOff, seq2Count);
        payloadOff += seq2Bytes;

        for (let i = 0; i < dictEntries; i++) {
          const parts = [];
          for (let j = 0; j < subChunksPerChunk; j++) {
            const idx = seq2Indices[i * subChunksPerChunk + j];
            if (idx >= dict2.length) {
              throw new Error(`Dict2 index out of range: ${idx} >= ${dict2.length}`);
            }
            parts.push(dict2[idx]);
          }
          const restored = Buffer.concat(parts).slice(0, chunkSize);
          this.dict.push(restored);
        }
      }
    } else {
      // ── 기존 복원 ──
      for (let i = 0; i < dictEntries; i++) {
        const chunk = Buffer.from(rawPayload.slice(payloadOff, payloadOff + chunkSize));
        this.dict.push(chunk);
        payloadOff += chunkSize;
      }
    }

    // Parse sequence section
    let data;
    if (seqCount > 0 && origBytes > 0) {
      const { values: indices } = varint.decodeArray(rawPayload, payloadOff, seqCount);

      // Restore chunks
      const chunks = [];
      for (const idx of indices) {
        if (idx >= this.dict.length) {
          throw new Error(`Dictionary index out of range: ${idx} >= ${this.dict.length}`);
        }
        chunks.push(this.dict[idx]);
      }

      const fullData = Buffer.concat(chunks);
      // Trim to origBytes (remove zero padding from last chunk)
      data = fullData.slice(0, origBytes);
    } else {
      data = Buffer.alloc(0);
    }

    return {
      data,
      bytesConsumed: off - startOffset,
    };
  }

  /**
   * 외부 딕셔너리 로드 (EXTERNAL_DICT 프레임 처리 시)
   */
  _loadExternalDict(chunkSize, baseDictCount) {
    // 이미 dict에 충분한 엔트리가 있으면 스킵 (멀티프레임 시)
    if (this.dict.length >= baseDictCount) return;

    const store = this._dictStore || new DictStore({
      dictDir: this._dictDir,
    });

    const loaded = store.load(chunkSize);

    if (loaded.entryCount < baseDictCount) {
      throw new Error(
        `외부 딕셔너리 엔트리 부족: ${loaded.entryCount} < ${baseDictCount}. ` +
        `dict-${chunkSize}.mshdict 파일이 손상되었거나 버전이 불일치합니다.`
      );
    }

    // 외부 딕셔너리 엔트리를 dict에 로드
    this.dict = loaded.dictChunks.slice(0, baseDictCount);
  }

  /**
   * Decompress payload
   */
  _decompress(data, codecId) {
    if (data.length === 0) return data;

    switch (codecId) {
      case CODEC.NONE:
        return data;
      case CODEC.GZIP:
        return zlib.gunzipSync(data);
      default:
        throw new Error(`Unsupported codec ID: ${codecId}`);
    }
  }
}

module.exports = { Unpacker };
